docs/rfc/0662-collaborative-objects.adoc

```diff
@@ -0,0 +1,789 @@
+to the network.
+
+== Implementation
+
+Every collaborative object is represented by a graph of automerge changes. See
+the <<appendix_automerge>> for more about automerge. We map this change graph
+to git as described in <<Change Commits>>, we store the tip of the change 
+graph under a `/cob/<typename>/<object ID>` reference, this allows us to 
+reconstruct the state of a collaborative object for a particular peer as 
+described in <<Reconstructing Collaborative Objects>>. This allows us to use
+git to replicate collaborative objects, see <<Fetching Collaborative Objects>>.
+
+We require some schema extensions to authenticate data in collaborative objects,
+these are described in <<Schema Extensions>>.
+
+
+=== Change Commits
+
+Given that automerge changes are a hash linked graph, we can map them to Git.
+We do so by wrapping each change in a commit. The commit points at a tree with
+the following layout
+
+[source]
+----
+.
+|--change
+|--salt
+|--manifest.toml
+----
+
+This tree contains a single change to a collaborative object. We will go into
+more details shortly. Any direct dependencies of this change are encoded in the 
+same manner and become the parents of this commit. This allows us to
+reconstruct the automerge depdency graph. 
+
+Along with the dependencies of the commit we also need to add the commit of the
+identity which created this commit. We need this identity to validate
+signatures and by making the commit a parent we ensure that git will replicate
+it for us. 
+
+A valid change commit must have three trailers:
+
+* `X-Rad-Signature`, as for identity documents
+* `X-Rad-Author-Parent`, this is the hash of the commit which references the
+  author identity. We use this trailer to avoid following the author commit
+  reference when constructing the automerge change graph
+* `X-Rad-Schema-Parent`, this is the hash of the parent commit which contains 
+  the schema of this object. See [schema commits](#schema-commits).
+
+
+==== `manifest.toml`
+
+The manifest is a TOML file containing some metadata about the object.
+Specifically it will contain:
+
+* `typename`, discussed above
+* `history_type`, always `"automerge"`, this is here to allow for different
+  CRDT implementations in future.
+
+
+Each object is also created with a JSON schema. The schema is represented by an
+initial `schema.json` and a series of schema migrations which extend that
+initial schema. Schema migrations will not be addressed in detail in this RFC
+but we will show their feasibility.
+
+
+==== `salt`
+
+This is a blob containing 64 random bytes to ensure that the hash of the tree
+of this object is unique. More details in <<Object IDs>>.
+
+==== `change`
+
+This is the automerge change which this commit introduces. It is a binary file
+which must contain a single change and it's dependents must be the dependents
+referenced by the parents of the commit.
+
+
+=== Object IDs
```

Comment:
The identities use the `tree` hash to enable a "multisig" scheme on the `commit` objects. The doc `blob` contains a hash link to the previous `tree`.

Iiuc, cobs could just be named after their initial `commit` (which contains sufficient entropy), or not? The histories of collaborators evolve independently anyways, and are merged only virtually.

Reply:
Yep using the initial `commit` works just as well, and probably better as it's authenticated. I was wary of doing this because it means we're depending on a bit more Git for the generation of the ID (tree hash + commit format), but on reflection I don't think being independent of Git is important for this.